src/state_machine.zig

```diff
@@ -1788,6 +1822,30 @@ pub fn StateMachineType(
             // checks.
             if (self.get_transfer(t.id)) |e| return create_transfer_exists(t, e);
 
+            if (t.flags.imported) {
+                // Allows past timestamp, but validates whether it regressed from the last
+                // inserted event.
+                // This validation must be called _after_ the idempotency checks so the user
+                // can still handle `exists` results when importing.
+                if (self.forest.grooves.transfers.objects.key_range) |*key_range| {
+                    if (t.timestamp <= key_range.key_max) {
+                        return .imported_event_timestamp_must_not_regress;
+                    }
+                }
+                if (t.timestamp < dr_account.timestamp) {
+                    return .imported_event_debit_account_must_not_advance;
+                }
+                if (t.timestamp < cr_account.timestamp) {
+                    return .imported_event_credit_account_must_not_advance;
+                }
+                if (t.timeout != 0) {
+                    assert(t.flags.pending);
+                    return .imported_event_timeout_must_be_zero;
+                }
+            }
+            assert(t.timestamp >= dr_account.timestamp);
+            assert(t.timestamp >= cr_account.timestamp);
```

Comment:
> ? In the absence of imports, there is a guarantee that timestamps are not shared between accounts and transfers. Why discard that invariant?

Sure! Let's keep it for non-imported transfers. 

However, it's an inconvenient special case: Why shouldn't we allow an imported Transfer whose timestamp is equal to one of the imported Accounts? I'm not saying we should allow it either!!!

Reply:
> Why shouldn't we allow an imported Transfer whose timestamp is equal to one of the imported Accounts?

Imo if we disallow it now, we can trivially allow it later if it turns out to be valuable.
But going the other direction would be much harder.

Reply:
Indeed, this is important: “to minimize surface area” (here meaning preserving stricter invariants)

Reply:
Sometimes with these things also, we might not yet know why we _should_ be strict, but later we may be grateful we were.... when it eventually becomes clear. i.e. Erring on the side of being strict closes potential semantic gaps down the line, and also preserves “optionality” (since we have more flexibility, we can decide to relax strict invariants, but one can't go the other way once relaxed).